### PR TITLE
Detect module support by testing its compilation

### PR DESCRIPTION
**Pull request by Mixaill**
Part of https://github.com/vitaut/modules/pull/2, but improved

* added function `modules_get_latest_cxx()` to get the latest standard version
* added function `modules_is_supported()` to build a simple module to determine actual module support
* use fallback sources in `add_module_library()` in case the check fails

### CODE REVIEW

File: modules.cmake [hunk 1]
```diff
@@ -43,25 +43,83 @@ if (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
...
+endfunction()
+
+
 # Adds a library compiled with C++20 module support.
 # `enabled` is a CMake variables that specifies if modules are enabled.
-# If modules are disabled `add_module_library` falls back to creating a
-# non-modular library.
 #
 # Usage:
-#   add_module_library(<name> [sources...] FALLBACK [sources...] [IF enabled])
```
Comment (vitaut) [inline modules.cmake]: `IF enabled` is needed by {fmt}, please keep it.

File: modules.cmake [hunk 2]
```diff
@@ -43,25 +43,83 @@ if (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
...
 
+
+# Receives latest available C++ standard version
+#
+# Usage:
+#   modules_get_latest_cxx(<variable_name>)
+#   if(<variable_name> GREATER 17)
+#      ...
+#   endif()
+function(modules_get_latest_cxx result)
```
Comment (vitaut) [inline modules.cmake]: Maybe modules_get_latest_cxx_std?

File: modules.cmake [hunk 3]
```diff
@@ -43,25 +43,83 @@ if (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
...
+    modules_get_latest_cxx(latest_standard)
+    if(latest_standard GREATER_EQUAL 20)
+
+      # create simple module file
+      set(temp_filepath "${CMAKE_CURRENT_BINARY_DIR}/module_test.cc")
+      file(WRITE "${temp_filepath}" "module;\nexport module module_test;\nexport void module_test_fun(){}")
+
+      # set needed compiler flags
+      set(compiler_flags "")
+      if (CMAKE_CXX_COMPILER_ID MATCHES "MSVC")
```
Comment (vitaut) [inline modules.cmake]: This can be just `if (MSVC)`.

File: modules.cmake [hunk 4]
```diff
@@ -43,25 +43,83 @@ if (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
...
+
+
+# Checks that compiler has support for C++ modules feature
+#
+# Usage:
+#   modules_is_supported(<variable_name>)
+#   if(<variable_name>)
+#      ...
+#   endif()
+function(modules_is_supported result)
```
Comment (vitaut) [inline modules.cmake]: is looks a bit weird with plural modules. Maybe just `modules_supported`?

File: modules.cmake [hunk 5]
```diff
@@ -43,25 +43,83 @@ if (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
...
+#   if(<variable_name> GREATER 17)
+#      ...
+#   endif()
+function(modules_get_latest_cxx result)
+  foreach(ver 26 23 20 17 14 11 98)
+    if(CMAKE_CXX_COMPILE_FEATURES MATCHES "cxx_std_${ver}")
+      set(${result} ${ver} PARENT_SCOPE)
+      break()
+    endif()
+  endforeach()
```
Comment (vitaut) [inline modules.cmake]: I think it would be better to find all occurrences of `cxx_std_*` and return the largest instead of trying different known standard versions which would need periodic update.

File: modules.cmake [hunk 6]
```diff
@@ -43,25 +43,83 @@ if (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
...
+#   endif()
+function(modules_is_supported result)
+    set(${result} FALSE PARENT_SCOPE)
+
+    # check standard version
+    modules_get_latest_cxx(latest_standard)
+    if(latest_standard GREATER_EQUAL 20)
+
+      # create simple module file
+      set(temp_filepath "${CMAKE_CURRENT_BINARY_DIR}/module_test.cc")
```
Comment (vitaut) [inline modules.cmake]: Where does CMake usually generate files like this?

**Review by vitaut — CHANGES_REQUESTED** [review]
Thanks for the PR!

File: modules.cmake (hunk 1 above)
Comment (Mixaill) [reply modules.cmake]: returned back

File: modules.cmake (hunk 2 above)
Comment (Mixaill) [reply modules.cmake]: renamed

File: modules.cmake (hunk 3 above)
Comment (Mixaill) [reply modules.cmake]: simplified

File: modules.cmake (hunk 4 above)
Comment (Mixaill) [reply modules.cmake]: renamed

File: modules.cmake (hunk 5 above)
Comment (Mixaill) [reply modules.cmake]: reimplemented

File: modules.cmake (hunk 6 above)
Comment (Mixaill) [reply modules.cmake]: Since CMake 3.25: `try_compile(SOURCE_FROM_VAR)`
Before CMake 3.25: `${CMAKE_BINARY_DIR}`

So, location changed and added `file(REMOVE ...)` to clean up stuff.

**Comment by Mixaill** [comment]
@vitaut notes resolved

File: modules.cmake [hunk 7]
```diff
@@ -57,11 +124,12 @@ function(add_module_library name)
   add_library(${name})
   set_target_properties(${name} PROPERTIES LINKER_LANGUAGE CXX)
 
-  if (NOT ${${AML_IF}})
-    # Create a non-modular library.
+  # Add fallback sources to the target in case modules are not supported or fallback was explicitly selected
+  modules_supported(modules_available)
+  if (NOT ${modules_available} OR (DEFINED AML_IF AND NOT AML_IF))
```
Comment (vitaut) [inline modules.cmake]: `AML_IF` should be checked first. This is used in {fmt} to avoid requiring both `FMT_MODULE` and setting the standard.

File: modules.cmake [hunk 8]
```diff
@@ -57,11 +124,12 @@ function(add_module_library name)
...
     target_sources(${name} PRIVATE ${AML_FALLBACK})
     return()
-  endif ()
+  endif()
```
Comment (vitaut) [inline modules.cmake]: nit: unnecessary change

File: modules.cmake [hunk 9]
```diff
@@ -43,6 +43,73 @@ if (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
...
+
+# Checks that compiler has support for C++ modules feature
+#
+# Usage:
+#   modules_is_supported(<variable_name>)
+#   if(<variable_name>)
+#      ...
+#   endif()
+function(modules_supported result)
+    set(${result} FALSE PARENT_SCOPE)
```
Comment (vitaut) [inline modules.cmake]: nit: extra indent

**Review by vitaut — CHANGES_REQUESTED** [review]

File: modules.cmake (hunk 8 above)
Comment (Mixaill) [reply modules.cmake]: reverted

File: modules.cmake (hunk 9 above)
Comment (Mixaill) [reply modules.cmake]: fixed

File: modules.cmake (hunk 7 above)
Comment (Mixaill) [reply modules.cmake]: fixed in the following way:

```
not defined -> use feature detection
defined     -> force the support to on/off
```

**Comment by Mixaill** [comment]
@vitaut issues resolved

**Comment by vitaut** [comment]
Thank you!
